### PR TITLE
[experiment] Compile skupper-router with g++ `-fpermissive`

### PR DESCRIPTION
**Pull request by jiridanek**
The bulk of the compilation problems (that `-fpermissive` does not suppress) is the different scoping in C and C++ regarding goto labels:

```
/home/jdanek/repos/skupper-router/tests/buffer_test.c: In function ‘char* test_buffer_field(void*)’:
/home/jdanek/repos/skupper-router/tests/buffer_test.c:366:1: error: jump to label ‘exit’
  366 | exit:
      | ^~~~
/home/jdanek/repos/skupper-router/tests/buffer_test.c:327:14: note:   from here
  327 |         goto exit;
      |              ^~~~
/home/jdanek/repos/skupper-router/tests/buffer_test.c:330:14: note:   crosses initialization of ‘uint32_t cmp_count’
  330 |     uint32_t cmp_count = 0;
      |              ^~~~~~~~~
```

Other issues (easy to suppress) are around C++ being more strict regarding `void *` values (that require explicit casts).

I have a list of hopeful directions that C++ compilation might enable, but so far no real results from that, except for some small enhancements that were found while doing this

* https://github.com/skupperproject/skupper-router/pull/249
* https://github.com/skupperproject/skupper-router/pull/248
* https://github.com/skupperproject/skupper-router/pull/247
* https://github.com/skupperproject/skupper-router/pull/211
* https://github.com/skupperproject/skupper-router/pull/82
* https://github.com/skupperproject/skupper-router/pull/1040
* https://github.com/skupperproject/skupper-router/pull/1039
* https://github.com/skupperproject/skupper-router/issues/1038

and 3rd party

* https://github.com/google/sanitizers/issues/1504

One thing I wanted to do with this was to experiment with datastructures (premade hashtables, etc, are more conveniently available in C++ than in C).

The PR does not seem to be too hard to keep up-to-date. I hope to do that for a while, and try something out with it.

Clang does not support -fpermissive.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,13 +18,13 @@
 ##
 
 cmake_minimum_required(VERSION 3.20)
-project(skupper-router C CXX)
```
Comment (jiridanek) [inline CMakeLists.txt:21]: I'd like to get this in main. Little CMakeLists.txt clarification.

File: CMakeLists.txt [hunk 2]
```diff
@@ -140,6 +146,7 @@ set(QPID_DISPATCH_CONFDIR ${SYSCONF_INSTALL_DIR}/skupper-router)
 # Set up runtime checks (valgrind, sanitizers etc.)
 include(cmake/RuntimeChecks.cmake)
 SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${SANITIZE_FLAGS}")
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${SANITIZE_FLAGS}")
```
Comment (jiridanek) [inline CMakeLists.txt:149]: Also something for main.

File: src/qd_asan_interface.h [hunk 3]
```diff
@@ -48,6 +48,10 @@
 
 #if QD_HAS_ADDRESS_SANITIZER
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
```
Comment (jiridanek) [inline src/qd_asan_interface.h:54]: Yet another thing for main.

File: src/router_core/connections.c [hunk 4]
```diff
@@ -1900,7 +1900,7 @@ static void qdr_link_inbound_second_attach_CT(qdr_core_t *core, qdr_action_t *ac
         // Handle incoming link cases
         //
         switch (link->link_type) {
-        case QD_LINK_ENDPOINT:
+        case QD_LINK_ENDPOINT: {
```
Comment (jiridanek) [inline src/router_core/connections.c:1903]: Writing body of a `case` into a block is IMO good practice. It's not necessary in C, however.

File: tests/lsan.supp [hunk 5]
```diff
@@ -9,16 +9,16 @@ leak:^IoAdapter_init$
 leak:^load_server_config$
 
 # to be triaged; system_tests_autolinks
-leak:^qdr_error_description$
+leak:^qdr_error_description
```
Comment (jiridanek) [inline tests/lsan.supp:12]: This is funny ASAN inconsistency. In C++, the suppression file entry is matched against function signature together with parameters, so the `$`s prevent a match.

File: src/python_embedded.c
Comment (jiridanek) [inline src/python_embedded.c:1]: Separate PR https://github.com/skupperproject/skupper-router/pull/1039

File: src/router_pynode.c
Comment (jiridanek) [inline src/router_pynode.c:1]: Separate PR https://github.com/skupperproject/skupper-router/pull/1039

File: CMakeLists.txt (hunk 1 above)
Comment (jiridanek) [reply CMakeLists.txt:21]: https://github.com/skupperproject/skupper-router/pull/1040

File: src/qd_asan_interface.h (hunk 3 above)
Comment (jiridanek) [reply src/qd_asan_interface.h]: Added to main in https://github.com/skupperproject/skupper-router/pull/497

File: CMakeLists.txt (hunk 2 above)
Comment (jiridanek) [reply CMakeLists.txt]: Added to main in https://github.com/skupperproject/skupper-router/pull/497

File: tests/lsan.supp (hunk 5 above)
Comment (jiridanek) [reply tests/lsan.supp]: Reported as https://github.com/google/sanitizers/issues/1504

**Comment by jiridanek** [comment]
This can probably be closed.

The `-fpermissive` option is g++-only, clang++ does not do this, and there appears to be absolutely zero future in making the router C code be in the C++-compatible subset of C, which was the idea of the experiment here.

There is a separate issue for using NULL/nullptr instead of 0 and this PR did not make any inroads in that aspect. So, if you agree, that is the first PR resolved right here.

* https://github.com/skupperproject/skupper-router/issues/245

**Comment by ganeshmurthy** [comment]
> This can probably be closed.

Closing this PR

**Pull request closed without merging**